src/arrow.js

```diff
@@ -0,0 +1,127 @@
+
+class Arrows extends Layer {
+  constructor(options) {
+    this.propertySpec = {
+      "arrow-min-size": {
+        type: "number",
+        minimum: 1,
+        default: 40,
+        expression: {
+          interpolated: true,
+          parameters: ["zoom"]
+        },
+        "property-type": "data-constant"
+      },
+      "particle-color": {
+        type: "color",
+        default: "white",
+        expression: {
+          interpolated: true,
+          parameters: ["zoom", "feature"]
+        },
+        "property-type": "data-driven"
+      },
+      "particle-halo-color": {
+        type: "color",
+        default: "rgba(0,0,0,0)",
+        expression: {
+          interpolated: true,
+          parameters: ["zoom"]
+        },
+        "property-type": "data-constant"
+      }
+    };
+    super(options);
+  }
+
+  initialize(map, gl) {
+    this.arrowsProgram = util.createProgram(gl, arrowsVert, arrowsFrag);
+    this.initializeGrid();
+  }
+
+  setParticleColor(expr) {
+    this.buildColorRamp(expr);
+  }
+
+  initializeGrid() {
+    this.cols = this.windData.width;
+    this.rows = this.windData.height;
+    const numTriangles = this.rows * this.cols * 2;
+    const numVertices = numTriangles * 3;
+    const positions = new Float32Array(2 * numVertices);
+    const corners = new Float32Array(2 * numVertices);
+    console.log(this.cols, this.rows);
+    for (let i = 0; i < this.cols; i++) {
+      for (let j = 0; j < this.rows; j++) {
+        const index = (i * this.rows + j) * 12;
+        positions.set([i, j, i, j, i, j, i, j, i, j, i, j], index);
+        corners.set([-1, 1, 1, 1, 1, -1, -1, 1, 1, -1, -1, -1], index);
+      }
+    }
+    this.positionsBuffer = util.createBuffer(this.gl, positions);
+    this.cornerBuffer = util.createBuffer(this.gl, corners);
+    console.log(positions, corners);
+  }
+
+  computeDimensions(gl, map) {
+    if (
+      map.getBounds().getEast() - 180 - (map.getBounds().getWest() + 180) >
+      0
+    ) {
+      return [gl.canvas.height, gl.canvas.height];
+    } else {
+      return [gl.canvas.width, gl.canvas.height];
+    }
+  }
+
+  draw(gl, matrix, dateLineOffset) {
+    const program = this.arrowsProgram;
+    gl.useProgram(program.program);
+
+    util.bindAttribute(gl, this.positionsBuffer, program.a_pos, 2);
+    util.bindAttribute(gl, this.cornerBuffer, program.a_corner, 2);
+
+    util.bindTexture(gl, this.windTexture, 0);
+    util.bindTexture(gl, this.colorRampTexture, 2);
+
+    gl.uniform1i(program.u_wind, 0);
+    gl.uniform1i(program.u_color_ramp, 2);
+
+    // compute downsampling
+    const [w, h] = this.computeDimensions(gl, this.map);
+    const z = this.map.getZoom();
+    const cols = Math.min(
+      Math.floor((Math.floor(z + 1) * w) / this.arrowMinSize),
+      this.cols
+    );
+    const rows = Math.min(
+      Math.floor((Math.floor(z + 1) * h) / this.arrowMinSize),
+      this.rows
+    );
+
+    gl.uniform2f(program.u_dimensions, cols, rows);
```

Comment:
Not sure what any of the code below does, I'm sure it is the actual drawing of arrows, maybe that should be encapsulated in a function called `drawArrow`, something that would give me some indicator of what it is really doing.